h2o-core/src/main/java/water/FrameSizeMonitor.java

```diff
@@ -0,0 +1,219 @@
+package water;
+
+import water.fvec.Frame;
+import water.fvec.NewChunk;
+import water.parser.FVecParseWriter;
+import water.util.Log;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+public class FrameSizeMonitor implements Runnable, Thread.UncaughtExceptionHandler {
+
+    private static final String LOG_LEVEL_PROP = "util.frameSizeMonitor.logLevel";
+    private static final String ENABLED_PROP = "util.frameSizeMonitor.enabled";
+    private static final String SAFE_COEF_PROP = "util.frameSizeMonitor.safetyCoefficient";
+    private static final String SAFE_FREE_MEM_DEFAULT_COEF = "0.2";
+
+    private static final int LOG_LEVEL;
+    private static final boolean ENABLED;
+    private static final float SAFE_FREE_MEM_COEF;
+    private static final int SLEEP_MS = 100;
+    private static final int MB = 1024 * 1024;
+    
+    private static final Map<Key<Job>, FrameSizeMonitor> registry = new HashMap<>();
+    
+    static {
+        LOG_LEVEL = Integer.parseInt(H2O.getSysProperty(LOG_LEVEL_PROP, String.valueOf(Log.DEBUG)));
+        ENABLED = H2O.getSysBoolProperty(ENABLED_PROP, false);
+        SAFE_FREE_MEM_COEF = Float.parseFloat(H2O.getSysProperty(SAFE_COEF_PROP, SAFE_FREE_MEM_DEFAULT_COEF));
+    }
+
+    private final Key<Job> jobKey;
+    private final Set<FVecParseWriter> writers = new HashSet<>();
+    private final long totalMemory = getTotalMemory();
+
+    private long committedMemory = 0;
+    
+    FrameSizeMonitor(Key<Job> jobKey) {
+        this.jobKey = jobKey;
+    }
+    
+    public static void get(Key<Job> jobKey, Consumer<FrameSizeMonitor> c) {
+        synchronized (registry) {
```

Comment:
Can we get away without using this global synchronized block?

IMHO you could just use a ConcurrentMap for registry and synchronize operation on the FrameSizeMonitor level.

Also with this mechanism disabled, do we need to enter the synchronized block at all?